core/runtime/setup.rs

```diff
@@ -0,0 +1,173 @@
+  let predictable_flags = "--predictable --random-seed=42";
+  let expose_natives_flags = "--expose_gc --allow_natives_syntax";
+
+  #[allow(clippy::useless_format)]
+  let flags = match (predictable, expose_natives) {
+    (false, false) => format!("{base_flags}"),
+    (true, false) => format!("{base_flags} {predictable_flags}"),
+    (false, true) => format!("{base_flags} {expose_natives_flags}"),
+    (true, true) => {
+      format!("{base_flags} {predictable_flags} {expose_natives_flags}")
+    }
+  };
+  v8::V8::set_flags_from_string(&flags);
+
+  let v8_platform = v8_platform.unwrap_or_else(|| {
+    if cfg!(any(test, feature = "unsafe_use_unprotected_platform")) {
+      // We want to use the unprotected platform for unit tests
+      v8::new_unprotected_default_platform(0, false)
+    } else {
+      v8::new_default_platform(0, false)
+    }
+    .make_shared()
+  });
+  v8::V8::initialize_platform(v8_platform.clone());
+  v8::V8::initialize();
+
+  v8::cppgc::initalize_process(v8_platform);
+}
+
+pub fn init_v8(
+  v8_platform: Option<v8::SharedRef<v8::Platform>>,
+  predictable: bool,
+  expose_natives: bool,
+) {
+  static DENO_INIT: Once = Once::new();
+  static DENO_PREDICTABLE: AtomicBool = AtomicBool::new(false);
+  static DENO_PREDICTABLE_SET: AtomicBool = AtomicBool::new(false);
+
+  if DENO_PREDICTABLE_SET.load(Ordering::SeqCst) {
+    let current = DENO_PREDICTABLE.load(Ordering::SeqCst);
+    assert_eq!(current, predictable, "V8 may only be initialized once in either snapshotting or non-snapshotting mode. Either snapshotting or non-snapshotting mode may be used in a single process, not both.");
+    DENO_PREDICTABLE_SET.store(true, Ordering::SeqCst);
+    DENO_PREDICTABLE.store(predictable, Ordering::SeqCst);
+  }
+
+  DENO_INIT
+    .call_once(move || v8_init(v8_platform, predictable, expose_natives));
+}
+
+pub fn init_cppgc(isolate: &mut v8::Isolate) -> v8::UniqueRef<v8::cppgc::Heap> {
+  let heap = v8::cppgc::Heap::create(
+    v8::V8::get_current_platform(),
+    v8::cppgc::HeapCreateParams::new(v8::cppgc::WrapperDescriptor::new(
+      0,
+      1,
+      crate::cppgc::DEFAULT_CPP_GC_EMBEDDER_ID,
+    )),
+  );
+
+  isolate.attach_cpp_heap(&heap);
+  heap
+}
+
+pub fn create_isolate_ptr() -> *mut v8::OwnedIsolate {
+  let align = std::mem::align_of::<usize>();
+  let layout = std::alloc::Layout::from_size_align(
+    std::mem::size_of::<*mut v8::OwnedIsolate>(),
+    align,
+  )
+  .unwrap();
+  assert!(layout.size() > 0);
+  let isolate_ptr: *mut v8::OwnedIsolate =
+    // SAFETY: we just asserted that layout has non-0 size.
+    unsafe { std::alloc::alloc(layout) as *mut _ };
+  isolate_ptr
+}
+
+pub fn create_isolate(
+  maybe_create_params: Option<v8::CreateParams>,
+  maybe_startup_snapshot: Option<Snapshot>,
+  will_snapshot: bool,
+  op_ctxs: &[OpCtx],
+  additional_references: &[v8::ExternalReference<'_>],
+) -> v8::OwnedIsolate {
+  let refs = bindings::external_references(op_ctxs, additional_references);
```

Comment:
If we pass the refs directly into this function it would decouple it from OpCtx.

Reply:
Already done, but not pushed :)